src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs

```diff
@@ -100,17 +108,80 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
-            password = ReadLineAsSecureString();
-            if (password == null)
+            if (confirmPassword)
             {
-                return null;
+                confirmPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ConfirmPassword, userName);
+                bool passwordsMatch;
+                do
+                {
+                    WriteToConsole(passwordPrompt, true);
+                    password = ReadLineAsSecureString();
+                    WriteToConsole(confirmPasswordPrompt, true);
+                    confirmedPassword = ReadLineAsSecureString();
+                    
+                    IntPtr passwordBstr = IntPtr.Zero;
+                    IntPtr confirmPasswordBstr = IntPtr.Zero;
+                    int match = 0;
+                    try
+                    {
+                        passwordBstr = Marshal.SecureStringToBSTR(password);
+                        confirmPasswordBstr = Marshal.SecureStringToBSTR(confirmedPassword);
+                        int passwordLength = Marshal.ReadInt32(passwordBstr, -4);
+                        int confirmPasswordLength = Marshal.ReadInt32(confirmPasswordBstr, -4);
+                        if (passwordLength == confirmPasswordLength)
+                        {
+                            for (int x = 0; x < passwordLength; ++x)
+                            {
+                                byte b1 = Marshal.ReadByte(passwordBstr, x);
+                                byte b2 = Marshal.ReadByte(confirmPasswordBstr, x);
+                                if (b1 != b2)
+                                {
+                                    // byte mismatch
+                                    match++;
+                                }
+                            }   
+                        }
+                        else
+                        {
+                            // length mismatch
+                            match++;
+                        }
+
+                        if (match == 0)
+                        {
+                            // passwords match
+                            passwordsMatch = true;
+                        }
+                        else
+                        {
+                            // passwords don't match
+                            WriteToConsole(StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PasswordMismatch), true);
+                            WriteLineToConsole();
+                            passwordsMatch = false;
+                        }
+                    }
+                    finally
+                    {
+                        if (confirmPasswordBstr != IntPtr.Zero)
+                            Marshal.ZeroFreeBSTR(confirmPasswordBstr);
+                        if (passwordBstr != IntPtr.Zero)
+                            Marshal.ZeroFreeBSTR(passwordBstr);
+                    }
+                } while (!passwordsMatch);
```

Comment:
Will Ctrl-C works?

Reply:
Yup, Ctrl-C works at any point in the process.